Add artifactory_certificate resource type

bodgit opened the pull request:
Reliant on atlassian/go-artifactory#17 so I can pass the PEM data as a string instead of a file path which sidesteps atlassian/go-artifactory#16 for now.

Example:
```hcl
resource "artifactory_certificate" "test" {
  alias   = "test"
  content = "${file("test.pem")}"
}
```

The magic is in the `CustomizeDiff` by computing the SHA256 fingerprint on the certificate in the config and comparing it to the fingerprint returned by the API on refresh. This way it can work out if someone changes the certificate outside of Terraform and forces an update.

Fixes #38

bodgit committed:
Add artifactory_certificate resource type

bodgit committed:
Docs

bodgit committed:
Tests

bodgit commented:
I've now added tests and documentation. I'm not sure if the tests I've added are actually running under Travis. If you cut a new release of `go-artifactory` then the go.mod should probably be tweaked again to use that instead of master.

bodgit commented:
@dillon-giacoppo Anything else needed here?

jamestoyer commented:
Hey @bodgit. @dillon-giacoppo has merged in a bunch of changes since this PR was created. Would it be possible for you to resolve them and check everything is working still? Either myself or Dillon will try to take a look then 🙂

bodgit commented:
@jamestoyer @dillon-giacoppo Done. I've also dropped the changes to `go.*` as `go-artifactory` v2.4.0 has the necessary changes. I also squashed some of the small bugfixes I made.

bodgit commented:
@jamestoyer @dillon-giacoppo Anyone? Bueller?

jamestoyer commented on pkg/artifactory/resource_artifactory_certificate.go:
```diff
@@ -0,0 +1,194 @@
+		},
+
+		CustomizeDiff: func(d *schema.ResourceDiff, v interface{}) error {
+			fingerprint, err := calculateFingerPrint(d.Get("content").(string))
```
There are a bunch of helpers in the base artifactory go package. You should probably use `getStringRef` instead

bodgit replied:
The code is using a `*schema.ResourceDiff` object, rather than a `*schema.ResourceData` object so there isn't currently a `getStringRef` method defined on that, (could do, however it's only going to be used for these two lines, not sure it's worth it?)

jamestoyer replied:
Ah cool, my bad

jamestoyer commented on pkg/artifactory/resource_artifactory_certificate.go:
```diff
@@ -0,0 +1,194 @@
+			if err != nil {
+				return err
+			}
+			if d.Get("fingerprint").(string) != fingerprint {
```
As above, use `getStringRef` instead

jamestoyer commented on pkg/artifactory/resource_artifactory_certificate.go:
```diff
@@ -0,0 +1,194 @@
+				return err
+			}
+			if d.Get("fingerprint").(string) != fingerprint {
+				d.SetNewComputed("fingerprint")
```
This can cause an error. It should probably be caught, logged and returned. [For example](https://github.com/atlassian/terraform-provider-artifactory/blob/master/pkg/artifactory/resource_artifactory_user.go#L90)

bodgit replied:
I've just checked the error and logged it rather than faff about with a `cascadingErr` for one potential error.

jamestoyer replied:
LGTM

jamestoyer commented on pkg/artifactory/resource_artifactory_certificate.go:
```diff
@@ -0,0 +1,194 @@
+		d.Set("issued_by", *cert.IssuedBy)
+		d.Set("issued_on", *cert.IssuedOn)
+		d.Set("issued_to", *cert.IssuedTo)
+		d.Set("valid_until", *cert.ValidUntil)
```
Could errors here be logged and returned too?

bodgit replied:
Done.

jamestoyer reviewed (changes requested):
@bodgit apologies for delay in review. I'm working on updates to the provider this week, so any changes made should be merged in this week and a new release cut

bodgit committed:
Address review comments

jamestoyer reviewed (changes requested):
Looks good. Could you do me a favour and merge in the latest upstream master changes? I've switched over to using Github actions and a new version of Golang, so want to make sure they all pass before merging

bodgit committed:
Merge remote-tracking branch 'upstream/master' into certificates

bodgit commented:
@jamestoyer done.
